src/box/box_trigger_def.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
```

Comment:
I would suggest to merge this entire file into box.h and turn `BOX_ON_RECOVERY_STATE` into a function. Separation into another file looks only inconsistent with other box-stuff. For example, `box_on_shutdown_trigger_list` is in box.h just fine. Same about the macro. It is not something perf-critical which needs to be inlined or macrosed.

If you made it a macro because of copying the strings, it should be fine to copy them on stack like:
```
const int max_size = 64;
char buf[max_size];
int len = strlen(state);
assert(len < max_size);
memcpy(buf, state, len + 1);
// run_triggers(buf);
```
Anyway we know all those strings and their max len.

Reply:
> I would suggest to merge this entire file into box.h

That's what I did initially, but it lead to introducing box.h dependency to memtx_engine.c
I wanted to avoid that. But I'll merge everything into box.h if you want me to.

> If you made it a macro because of copying the strings, it should be fine to copy them on stack like:

Yep, I didn't want to make the copy, since the string's copied when passed to lua anyway. This would leave us with two copies of the same string.
Anyway, I'm fine with turning this to a function:
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 4a3339727..66f8c49a3 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -416,7 +416,7 @@ box_do_set_orphan(bool orphan)
 	is_orphan = orphan;
 	box_update_ro_summary();
 	if (!is_orphan)
-		BOX_ON_RECOVERY_STATE("synced");
+		box_run_on_recovery_state("synced");
 }
 
 void
@@ -3772,7 +3772,7 @@ bootstrap_master(const struct tt_uuid *replicaset_uuid)
 	if (gc_checkpoint() != 0)
 		panic("failed to create a checkpoint");
 
-	BOX_ON_RECOVERY_STATE("snapshot_recovered");
+	box_run_on_recovery_state("snapshot_recovered");
 }
 
 /**
@@ -3822,7 +3822,7 @@ bootstrap_from_master(struct replica *master)
 					APPLIER_FINAL_JOIN;
 	applier_resume_to_state(applier, wait_state, TIMEOUT_INFINITY);
 
-	BOX_ON_RECOVERY_STATE("snapshot_recovered");
+	box_run_on_recovery_state("snapshot_recovered");
 
 	/*
 	 * Process final data (WALs).
@@ -4054,7 +4054,7 @@ local_recovery(const struct tt_uuid *instance_uuid,
 	 */
 	memtx_engine_recover_snapshot_xc(memtx, checkpoint_vclock);
 
-	BOX_ON_RECOVERY_STATE("snapshot_recovered");
+	box_run_on_recovery_state("snapshot_recovered");
 
 	engine_begin_final_recovery_xc();
 	recover_remaining_wals(recovery, &wal_stream.base, NULL, false);
@@ -4358,7 +4358,7 @@ box_cfg_xc(void)
 
 	rmean_cleanup(rmean_box);
 
-	BOX_ON_RECOVERY_STATE("wal_recovered");
+	box_run_on_recovery_state("wal_recovered");
 
 	/* Follow replica */
 	replicaset_follow();
@@ -4399,8 +4399,8 @@ box_cfg_xc(void)
 	say_info("ready to accept requests");
 
 	if (is_bootstrap_leader) {
-		BOX_ON_RECOVERY_STATE("indexes_built");
-		BOX_ON_RECOVERY_STATE("synced");
+		box_run_on_recovery_state("indexes_built");
+		box_run_on_recovery_state("synced");
 	}
 	if (!is_bootstrap_leader) {
 		replicaset_sync();
diff --git a/src/box/box_trigger_def.h b/src/box/box_trigger_def.h
index 558890cf1..8bd90ad00 100644
--- a/src/box/box_trigger_def.h
+++ b/src/box/box_trigger_def.h
@@ -12,20 +12,24 @@
  */
 extern struct rlist box_on_recovery_state;
 
-/*
- * The macro is intended to simplify trigger invocation.
- *
- * on_recovery_state triggers expect a string, but passing a string literal to
- * them turns out quite ugly, because trigger_run takes void *. So one ends up
- * with trigger_run(trig, (char *)"string_literal");
- *
- * Compound literals solve this problem in a less ugly way, besides they really
- * can be written to: trigger_run(trig, (char[]){"string_literal"}). But
- * unfortunately they aren't supported in c++.
- *
- * So we are left with this:
- */
-#define BOX_ON_RECOVERY_STATE(str) ({\
-	char arr[] = str;\
-	trigger_run(&box_on_recovery_state, arr);\
-})
+/** Run box_on_recovery_state triggers with state represented by \a str. */
+static inline int
+box_run_on_recovery_state(const char *str)
+{
+	/*
+	 * on_recovery_state triggers expect a string, but passing a string
+	 * literal to them turns out quite ugly, because trigger_run takes
+	 * void *. So one ends up with
+	 * trigger_run(trig, (char *)"string_literal");
+	 *
+	 * Compound literals solve this problem in a less ugly way, besides they
+	 * can really be written to:
+	 * trigger_run(trig, (char[]){"string_literal"}).
+	 * But unfortunately they aren't supported in c++.
+	 * So we are left with this:
+	 */
+	char buf[64];
+	size_t len = strlcpy(buf, str, sizeof(buf));
+	assert(len < sizeof(buf));
+	return trigger_run(&box_on_recovery_state, buf);
+}
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 554094fcf..4049e3848 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -437,7 +437,7 @@ memtx_engine_begin_final_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		BOX_ON_RECOVERY_STATE("indexes_built");
+		box_run_on_recovery_state("indexes_built");
 	}
 	return 0;
 }
@@ -456,7 +456,7 @@ memtx_engine_begin_hot_standby(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		BOX_ON_RECOVERY_STATE("indexes_built");
+		box_run_on_recovery_state("indexes_built");
 	}
 	return 0;
 }
@@ -476,7 +476,7 @@ memtx_engine_end_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		BOX_ON_RECOVERY_STATE("indexes_built");
+		box_run_on_recovery_state("indexes_built");
 	}
 	xdir_collect_inprogress(&memtx->snap_dir);

```

</details>

Reply:
I wouldn't worry about the header dependency. It depends on libbox anyway, whatever you do. Lets better do it without new files.

About the copy - your current solution also does the copying on the stack. There wouldn't be copying if you would store it as const char * (I understand why you can't do it here), but when you do char[] the string literal is copied onto the stack.

Reply:
> I wouldn't worry about the header dependency. It depends on libbox anyway, whatever you do. Lets better do it without new files.

Ok, I moved everything to `box.h` / `box.cc`.

> About the copy - your current solution also does the copying on the stack. There wouldn't be copying if you would store it as const char * (I understand why you can't do it here), but when you do char[] the string literal is copied onto the stack.

Yep, makes sense. It was just hidden.